### ISSUE TITLE
Tests for Exercise 2.2.9

### ISSUE THREAD

**Issue by prathyvsh** [kesava/eopl#1]
Hey, great job starting with the first edition!
Have heard that three editions are three distinct books.
Yet to get into 2nd and 3rd editions but I was working on Chapter 2 of the first one.
I came across the tough cookie that is 2.2.9 and was comparing my results with yours.

These tests on yours seem to give wrong answer on [Exercise 2.2.9](https://github.com/kesava/eopl/blob/master/first-edition/chap2.scm#L447):

```
> (car&cdr 'a '() 'fail)
'(lambda (x) ())
> (car&cdr 'a '(() ((a))) 'fail)
'(lambda (x) (car (car (car ()))))
> (car&cdr 'a '(() (a)) 'fail)
'(lambda (x) (car (car ())))
> (car&cdr 'a '(() (b) a) 'fail)
'(lambda (x) (car (cdr ())))
> (car&cdr 'a '(b c d () e (a)) 'fail)
'(lambda (x) (car (car (cdr (cdr cdr)))))
> (car&cdr 'a '(() () a) 'fail)
. . car: contract violation
  expected: pair?
  given: '()
```

**Comment by kesava** [comment]
yeah, this was a tough one. Was stuck on it for a few days. i later realized my solution is buggy, but felt too exhausted to go back and fix cases outside the book. Needs a proper stack inside.

**Comment by prathyvsh** [comment]
Think once you properly take care of each of syntactic category of los and lst, I think the solution falls out somewhat easily. I used the idea of getting the path of the Binary Search Tree and then reversing it and then nesting it, which I think covers all of the cases.

If you use tail recursion instead of structural one, the need for reversal is eliminated which makes it a natural solution. Think its easily fixable as you have got most of the idea of arriving it under control.
